src/compiler/checker.ts

```diff
@@ -11994,7 +11994,7 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             // to preserve this type. In fact, we need to _prefer_ that type, but it won't
             // be assigned until contextual typing is complete, so we need to defer in
             // cases where contextual typing may take place.
-            if (!links.type && !isParameterOfContextSensitiveSignature(symbol) && !checkMode) {
+            if (!links.type && !isParameterOfContextSensitiveSignature(symbol) && !(checkMode & ~CheckMode.TypeOnly)) {

```

Comment:
this cache suppression was introduced in https://github.com/microsoft/TypeScript/pull/56753 and its original PR https://github.com/microsoft/TypeScript/pull/50586 was first created before `CheckMode.TypeOnly` was even a thing (that was introduced in https://github.com/microsoft/TypeScript/pull/54380 )

It would be an easy thing to miss either way but partially the bug was born because that PR was long-lived and a new concept was introduced to the checker in the meantime.